src/main/java/picard/fastq/SamToBfqWriter.java

```diff
@@ -0,0 +1,447 @@
+ * THE SOFTWARE.
+ */
+package picard.fastq;
+
+import htsjdk.samtools.ReservedTagConstants;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.filter.AggregateFilter;
+import htsjdk.samtools.filter.FailsVendorReadQualityFilter;
+import htsjdk.samtools.filter.FilteringSamIterator;
+import htsjdk.samtools.filter.SamRecordFilter;
+import htsjdk.samtools.filter.TagFilter;
+import htsjdk.samtools.filter.WholeReadClippedFilter;
+import htsjdk.samtools.util.BinaryCodec;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.samtools.util.Log;
+import htsjdk.samtools.util.PeekableIterator;
+import htsjdk.utils.ValidationUtils;
+import picard.PicardException;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Class to take unmapped reads in SAM/BAM/CRAM file format and create Maq binary fastq format file(s) --
+ * one or two of them, depending on whether it's a paired-end read.  This relies on the unmapped
+ * file having all paired reads together in order.
+ */
+public class SamToBfqWriter {
+
+    private static final int SEED_REGION_LENGTH = 28;
+    private static final int MAX_SEED_REGION_NOCALL_FIXES = 2;
+
+    private final File bamFile;
+    private final String outputPrefix;
+    private final String namePrefix;
+    private final int nameTrim;
+    private final File referenceSequence;
+    private final boolean pairedReads;
+    private int wrote = 0;
+    private int increment = 1;
+    private int chunk = 0;
+    private BinaryCodec codec1;
+    private BinaryCodec codec2;
+    private final Log log = Log.getInstance(SamToBfqWriter.class);
+    private final boolean includeNonPfReads;
+    private final boolean clipAdapters;
+    private final Integer basesToWrite;
+
+    /**
+     * Constructor
+     *
+     * @param bamFile        the BAM file to read from
+     * @param outputPrefix   the directory and file prefix for the binary fastq files
+     * @param total          the total number of records that should be written, drawn evenly
+     *                       from throughout the file (null for all).
+     * @param chunk          the maximum number of records that should be written to any one file
+     * @param pairedReads    whether these reads are from  a paired-end run
+     * @param namePrefix     The string to be stripped off the read name
+     *                       before writing to the bfq file. May be null, in which case
+     *                       the name will not be trimmed.
+     * @param includeNonPfReads whether to include non pf-reads
+     * @param clipAdapters    whether to replace adapters as marked with XT:i clipping position attribute
+     */
+    public SamToBfqWriter(final File bamFile, final File referenceSequence, final String outputPrefix, final Integer total,
```

Comment:
Should this be `File samFile`? (and elsewhere)

Reply:
sure. elsewhere, in this file? or elsewhere in other files?